File: src/leetcode/p0100_0199/p0108_convert_sorted_array_to_binary_search_tree/ConvertSortedArrayToBinarySearchTree.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.treenode;

import common.provide.TreeNode;

public interface ConvertSortedArrayToBinarySearchTree {
    /**
     * Given an integer array nums where the elements are sorted in
     * ascending order, convert it to a height-balanced binary search
     * tree.
     *
     * A height-balanced binary tree is a binary tree in which the
     * depth of the two subtrees of every node never differs by more
     * than one.
     *
     * @param nums 1 <= nums.length <= 10^4
     *             -10^4 <= nums[i] <= 10^4
     *             is sorted in a strictly increasing order
     * @return height-balanced Binary Search Tree
     */
    TreeNode sortedArrayToBST(int[] nums);
}

File: src/leetcode/p0100_0199/p0108_convert_sorted_array_to_binary_search_tree/convert_sorted_array_to_binary_search_tree.py
from abc import ABC, abstractmethod
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class ConvertSortedArrayToBinarySearchTree(ABC):
    """
    Given an integer array nums where the elements are sorted in ascending order, convert it to a binary search tree.
    """
    @abstractmethod
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        """
        Args:
            nums:
                - 1 <= nums.length <= 10^4
                - -10^4 <= nums[i] <= 10^4
                - nums is sorted in a strictly increasing order.
        """
        pass

class ConvertSortedArrayToBinarySearchTreeRecursive(ConvertSortedArrayToBinarySearchTree):
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def recursive(left, right):
            if left > right:
                return None
            mid = (left + right) // 2
            root = TreeNode(nums[mid])
            root.left = build(left, mid - 1)
            root.right = build(mid + 1, right)
            return root

        return recursive(0, len(nums) - 1)
